Add option to print all paths as absolute paths

carlos-granados opened the pull request:
Currently we have several places in our code where we handle paths to relativize them. This PR replaces them with a single service that can print paths and which is configurable. And we add a new option which can be used to tell Behat to print all paths as absolute instead of relative paths. This option is available on the command line and also in the Php and Yaml configs

carlos-granados commented on features/bootstrap/FeatureContext.php:
```diff
@@ -457,6 +457,24 @@ public function theOutputShouldContain(PyStringNode $text)
+     *
+     * @param PyStringNode $text PyString text instance
+     */
+    public function theOutputWithAbsolutePathsShouldContain(PyStringNode $text)
```
We add a new step that allows us to check the output without filtering out the absolute paths and instead replacing them with a placeholder

carlos-granados commented on features/convert_config.feature:
```diff
@@ -400,6 +400,24 @@ Feature: Convert config
       """
     And the temp "formatters.yaml" file should have been removed
 
+  Scenario: path options
```
We test that the yaml config is correctly converted to the PHP config for this new option

carlos-granados commented on phpstan.dist.neon:
```diff
@@ -13,6 +13,8 @@ parameters:
                 - src/Behat/Behat/Definition/Printer/ConsoleDefinitionPrinter.php
                 - src/Behat/Behat/EventDispatcher/Cli/StopOnFailureController.php
                 - src/Behat/Behat/Output/Node/Printer/ListPrinter.php
+                - src/Behat/Behat/Output/Node/Printer/Pretty/PrettyPathPrinter.php
```
Now we are not using some constructor arguments in some classes but we can't remove them so we tell PhpStan to ignore them

carlos-granados commented on src/Behat/Behat/ApplicationFactory.php:
```diff
@@ -83,6 +84,7 @@ protected function getDefaultExtensions()
             new AutoloaderExtension(['' => '%paths.base%/features/bootstrap']),
             new SuiteExtension($processor),
             new OutputExtension('pretty', $this->getDefaultFormatterFactories($processor), $processor),
+            new PathOptionsExtension(),
```
Add a new extension that adds the new path options

carlos-granados commented on src/Behat/Behat/Output/Node/Printer/ListPrinter.php:
```diff
@@ -45,24 +46,26 @@ final class ListPrinter
         ExceptionPresenter $exceptionPresenter,
         TranslatorInterface $translator,
-        $basePath,
+        string $basePath,
```
This parameter is not used any more because the configurable path printer service takes care of handling path printing

carlos-granados commented on src/Behat/Behat/Output/Node/Printer/ListPrinter.php:
```diff
@@ -45,24 +46,26 @@ final class ListPrinter
-        $this->basePath = $basePath;
+    }
+
+    public function setConfigurablePathPrinter(ConfigurablePathPrinter $configurablePathPrinter)
```
We add a function to be able to set the new configurable path printer service as we cannot inject it through the constructor to keep BC

carlos-granados commented on src/Behat/Behat/Output/Node/Printer/ListPrinter.php:
```diff
@@ -84,7 +87,7 @@ public function printScenariosList(OutputPrinter $printer, $intro, $resultCode,
         $printer->writeln(sprintf('--- {+%s}%s{-%s}' . PHP_EOL, $style, $intro, $style));
         foreach ($scenarioStats as $stat) {
-            $path = $this->relativizePaths((string) $stat);
+            $path = $this->configurablePathPrinter->processPathsInText((string) $stat);
```
All paths are now printed using the new service

carlos-granados commented on src/Behat/Behat/Output/Node/Printer/ListPrinter.php:
```diff
@@ -281,35 +284,19 @@ private function printStepStat(
         $printer->writeln();
     }
 
-    /**
```
And now this function is not needed

carlos-granados commented on src/Behat/Behat/Output/ServiceContainer/Formatter/PrettyFormatterFactory.php:
```diff
@@ -220,6 +220,7 @@ protected function loadCorePrinters(ContainerBuilder $container)
             new Reference('output.node.printer.pretty.width_calculator'),
             '%paths.base%',
         ]);
+        $definition->addMethodCall(('setConfigurablePathPrinter'), [new Reference('configurable.path.printer')]);
```
We inject the new service into the classes that need it

carlos-granados commented on src/Behat/Config/Profile.php:
```diff
@@ -102,6 +106,13 @@ public function withPrintUnusedDefinitions(bool $printUnusedDefinitions = true):
         return $this;
     }
 
+    public function withPathOptions(bool $printAbsolutePaths): self
```
New config option to set this option using PHP

carlos-granados commented on src/Behat/Config/Profile.php:
```diff
@@ -215,6 +227,22 @@ private function addUnusedDefinitionsToExpr(Expr &$expr): void
         }
     }
 
+    private function addPathOptionsToExpr(Expr &$expr): void
```
And new yaml to php converter for this option

carlos-granados commented on src/Behat/Testwork/PathOptions/Printer/ConfigurablePathPrinter.php:
```diff
@@ -0,0 +1,46 @@
+
+namespace Behat\Testwork\PathOptions\Printer;
+
+final class ConfigurablePathPrinter
```
This is the new service that prints paths, adding an option to decide if they should be printed as relative (default) or absolute

acoulton reviewed (commented):
This will be great @carlos-granados, thanks. Just a couple of thoughts.

acoulton commented:
I think this will also help with #1601

carlos-granados commented on features/bootstrap/FeatureContext.php:
```diff
@@ -569,7 +587,11 @@ private function getOutput()
+            realpath(dirname(dirname(__DIR__))) . DIRECTORY_SEPARATOR,
+            $removeAbsolutePaths ? '' : 'BASE_PATH/',
+            $output
+        );
```
Yes, we can always replace with BASE_PATH, there was one place were I've had to be explicit about this. I also updated the placeholder to something that works on Windows

carlos-granados commented on features/print_absolute_paths.feature:
```diff
@@ -0,0 +1,58 @@
+      --- Failed scenarios:
+
+          BASE_PATH/tests/Fixtures/PrintAbsolutePaths/features/test.feature:3
+      """
```
I started working on this feature before I finished the config converter and this is why this was here. Now that we have the config converter I agree that we can remove it.

carlos-granados commented:
@acoulton regarding your comment about the constructors, agree that your suggestion might be a better idea, updated. I also added a constant for the id of the configurable path printer service

acoulton reviewed (commented):
Great, thanks @carlos-granados - just one tiny thing I spotted in the diff.
